### PR TITLE
Re #25031. Allow for building workbench with only Qt5 installed.

### LINKED ISSUE #25031: Only require Qt5 (no Qt4) when building just the workbench (not MantidPlot)
**Original reporter:** @marshallmcdonnell 

### Expected behavior

Be able to build the workbench and not MantidPlot with only Qt5 installed (no Qt4 requirement). Thus, building with something along the lines of:
```
cmake -DENABLE_MANTIDPLOT=FALSE -DENABLE_WORKBENCH=TRUE /path/to/src
```
### Actual behavior
There is still Qt4 dependence in the CMakeFiles for just building the workbench. Needs to have if-else statements for workbench or mantidplot in these files.

### Steps to reproduce the behavior

Uninstall Qt4 libraries, try to build with pseudo-cmake command above. Should trip on trying to find Qt4 (Qt4, QScintilla w/ Qt4 bindings, Qwt, etc.)

### Platforms affected
All...

### PR DESCRIPTION
**Pull request by marshallmcdonnell**
**Description of work.**

This allows for building the workbench  with just Qt5 installed and without Qt4 (ie `-DENABLE_MANTIDPLOT=FALSE -DENABLE_WORKBENCH=TRUE`).

**My testing for Qt5 only installation (using containers):**
I used docker to create Ubuntu 18.04 and CentOS7 images with Qt4 removed. Dockerfiles to build the images are here (`*.txt` file extension to allow for uploading):

[UbuntuBionicTesting.Dockerfile.txt](https://github.com/mantidproject/mantid/files/2933357/UbuntuBionicTesting.Dockerfile.txt)
[CentOS7Testing.Dockerfile.txt](https://github.com/mantidproject/mantid/files/2933359/CentOS7Testing.Dockerfile.txt)

Then built the images without Qt4 installed via:
```
docker build --file=UbuntuBionicTesting.Dockerfile.txt --tag mantidproject/ubuntubionic-testing:qt5-workbench

docker build --file=CentOS7Testing.Dockerfile.txt --tag mantidproject/centos7-testing:qt5-workbench
```

Then built Mantid in containers based off the images by mounting the mantid source and build directories (only showing Ubuntu version):
```
export HOST_MANTID_SRC=/path/to/mantid/src
mkdir $HOST_MANTID_SRC/build-no_qt4_workbench_install
docker run --rm -it \
           --env="DISPLAY" \
           --env="QT_X11_NO_MITSHM=1" \
           --volume="/tmp/.X11-unix:/tmp/.X11-unix:rw" \
           --volume="$HOST_MANTID_SRC/build-no_qt4_workbench_install:/mantid_build" \
           --volume="$HOST_MANTID_SRC:/mantid_src" \
           mantidproject/ubuntubionic-testing:qt5-workbench
```
Then, built workbench only:
```
cd /mantid_build
cmake -GNinja -DENABLE_MANTIDPLOT=FALSE -DENABLE_WORKBENCH=TRUE /mantid_src
```
And launched to ensure the workbench opens:
```
/mantid_build/bin/launch_mantidworkbench.sh
```
NOTE: I used the following "unsecure, lazy, reckless" way to get my X11 stuff to work using the following based on the examples [here](http://wiki.ros.org/docker/Tutorials/GUI):
```
xhost +local:root # for the lazy and reckless
<run "docker run ..." command and launch workbench>
xhost -local:root
```

**To test for Qt5 only installation (using machine):**
I didn't do this but it would require uninstalling Qt4. Example for Debian:
```
sudo apt-get remove -y \
  qt4-default \
  qt4-qmake \
  qt4-dev-tools
```

Then install with `-DENABLE_MANTIDPLOT=FALSE -DENABLE_WORKBENCH=TRUE` flags and make sure workbench builds and works.

**My testing for Qt4 and Qt5 installed (using machine):**
Simply build normally (with `-DENABLE_MANTIDPLOT=TRUE -DENABLE_WORKBENCH=TRUE`) and ensure both MantidPlot and workbench still work.

**Further testing**
I **have not tested this for Mac or Windows machines** so those would be necessary tests before approval.

Fixes #25031 

**Release Note**
Added for v4.0.0

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/IndividualTicketTesting/)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by marshallmcdonnell** [comment]
Had to add names for Qt5 Qscintilla2 for conda library. For my conda environment with qscintilla install, I have the following:

**Review by peterfpeterson — APPROVED** [review]
I verified this worked by configuring using
```
scl enable devtoolset-7 "cmake3 -GNinja -DENABLE_WORKBENCH=on -DENABLE_MANTIDPLOT=off -DENABLE_CPACK=on /home/pf9/code/mantid"
```
Then I installed the package, started the workbench (`launch_mantidplot.sh` was correctly missing), then loading a workspace and showing the instrument view.

File: qt/widgets/common/CMakeLists.txt [hunk 1]
```diff
@@ -734,67 +734,71 @@ if ( ENABLE_WORKBENCH )
...
-    Qt4::Qscintilla
-  INSTALL_DIR
-    ${LIB_DIR}
-  OSX_INSTALL_RPATH
-    @loader_path/../MacOS
-    @loader_path/../Libraries
-  LINUX_INSTALL_RPATH
-    "\$ORIGIN/../${LIB_DIR}"
-)
+if ( ENABLE_MANTIDPLOT )
```
Comment (martyngigg) [inline qt/widgets/common/CMakeLists.txt]: Are these `if` blocks necessary. I thought `mtd_add_qt_library` was setup to skip Qt4  targets if `ENABLE_MANTIDPLOT=FALSE` [QtTargetFunctions.cmake](https://github.com/mantidproject/mantid/blob/master/buildconfig/CMake/QtTargetFunctions.cmake#L23)? The idea was to reduce the number of places we would need this check.

File: qt/widgets/common/CMakeLists.txt [hunk 2]
```diff
@@ -839,29 +843,31 @@ set( TEST_FILES
   test/WorkspacePresenter/WorkspacePresenterTest.h
 )
 
-mtd_add_qt_tests (TARGET_NAME MantidQtWidgetsCommonTest
```
Comment (martyngigg) [inline qt/widgets/common/CMakeLists.txt]: Similar comment to above.

File: qt/widgets/common/CMakeLists.txt [hunk 3]
```diff
@@ -904,11 +912,20 @@ mtd_add_qt_tests (TARGET_NAME MantidQtWidgetsCommonTest
...
+if ( WIN32 AND ENABLE_MANTIDPLOT)
   set ( QT_CONF qt.conf )
   configure_file ( qt.conf.template ${CMAKE_CURRENT_BINARY_DIR}/${QT_CONF} @ONLY )
   add_custom_command ( TARGET MantidQtWidgetsCommonQt4 POST_BUILD
                        COMMAND ${CMAKE_COMMAND} ARGS -E copy_if_different
                                ${CMAKE_CURRENT_BINARY_DIR}/${QT_CONF}
                                ${CMAKE_LIBRARY_OUTPUT_DIRECTORY}/${CMAKE_CFG_INTDIR}/${QT_CONF} )
 endif()
+
+if ( WIN32 AND ENABLE_WORKBENCH)
```
Comment (martyngigg) [inline qt/widgets/common/CMakeLists.txt]: Does this work/Is it necessary? We didn't use `qt.conf` before now as Qt only picks it up from the application executable directory and for us this is now Python and not our application.

**Review by martyngigg — CHANGES_REQUESTED** [review]
Thanks for looking at this. I've just a few questions.

File: qt/widgets/common/CMakeLists.txt (hunk 3 above)
Comment (marshallmcdonnell) [reply qt/widgets/common/CMakeLists.txt]: Sorry, don't quite know (works, necessary, etc.).

Should I just leave it as `if (WIN32)` and remove the separate flags for workbench and mantidplot? Just not sure and don't have a WIN32 system to check this on.

File: qt/widgets/common/CMakeLists.txt (hunk 1 above)
Comment (marshallmcdonnell) [reply qt/widgets/common/CMakeLists.txt]: Oh, good call, didn't catch that. Removing and testing now.

**Comment by marshallmcdonnell** [comment]
Hopefully, [this PR](https://github.com/mantidproject/mantid/pull/25062) will get merged rather quickly and I'll resolve conflicts with my release notes here.

File: docs/source/release/index.rst [hunk 4]
```diff
@@ -13,6 +13,7 @@ Release Notes
 
 
 * :ref:`v4.0.0 <v4.0.0>`
+* :ref:`v4.1.0 <v4.1.0>`
```
Comment (martyngigg) [inline docs/source/release/index.rst]: I think we need to remove this line too?

File: qt/widgets/common/CMakeLists.txt [hunk 5]
```diff
@@ -839,29 +839,31 @@ set( TEST_FILES
...
-    DataObjects
-    ${GMOCK_LIBRARIES}
-    ${GTEST_LIBRARIES}
-  MTD_QT_LINK_LIBS
-    MantidQtWidgetsCommon
-  PARENT_DEPENDENCIES
-    GUITests
-)
+if ( ENABLE_MANTIDPLOT )
+  mtd_add_qt_tests (TARGET_NAME MantidQtWidgetsCommonTest
```
Comment (martyngigg) [inline qt/widgets/common/CMakeLists.txt]: Is this `if` required? The `mtd_add_qt_tests` function is suposed to operate in a similar manner to the library target and will auto-disable the Qt4 tests if `ENABLE_MANTIDPLOT=False`.

**Review by martyngigg — CHANGES_REQUESTED** [review]

File: docs/source/release/index.rst (hunk 4 above)
Comment (marshallmcdonnell) [reply docs/source/release/index.rst]: Good catch, thanks, removed.

File: qt/widgets/common/CMakeLists.txt (hunk 5 above)
Comment (marshallmcdonnell) [reply qt/widgets/common/CMakeLists.txt]: No, its not, thanks. I checked the `mtd_add_qt_<suffix> suffix={tests, library, executable}` and think I have fixed these all now to remove unnecessary if-statments.

**Review by martyngigg — APPROVED** [review]
This looks good to me now.

File: qt/widgets/common/CMakeLists.txt (hunk 5 above)
Comment (martyngigg) [reply qt/widgets/common/CMakeLists.txt]: Thanks :+1:

**Pull request merged**